File: Leetcode-one-by-one/src/Little Shino and K Ancestor.swift
import Foundation

let MAXN = 1000001

var a = Array(repeating: [Int](), count: MAXN)
var c = Array(repeating: [Int](), count: MAXN)
var v = Array(repeating: 0, count: MAXN)
var b = Array(repeating: 0, count: MAXN)
var d = Array(repeating: 0, count: MAXN)

func dfs(_ s: Int, _ k: Int) {
    v[s] = 1

    if c[b[s]].count >= k {
        let size = c[b[s]].count
        d[s] = c[b[s]][size - k]
    } else {
        d[s] = -1
    }

    c[b[s]].append(s)

    for nei in a[s] {
        if v[nei] == 0 {
            dfs(nei, k)
        }
    }

    if !c[b[s]].isEmpty {
        c[b[s]].removeLast()
    }
}

// Input
let input = readLine()!.split(separator: " ").map { Int($0)! }
let n = input[0], k = input[1]

let values = readLine()!.split(separator: " ").map { Int($0)! }
for i in 1...n {
    b[i] = values[i - 1]
}

for _ in 1..<n {
    let edge = readLine()!.split(separator: " ").map { Int($0)! }
    let x = edge[0], y = edge[1]
    a[x].append(y)
    a[y].append(x)
}

dfs(1, k)

for i in 1...n {
    print(d[i], terminator: " ")
}

File: Leetcode-one-by-one/src/Little Shino and K Ancestor.java
import java.util.*;

public class Main {
    static List<Integer>[] a = new ArrayList[1000001];
    static List<Integer>[] c = new ArrayList[1000001];
    static int[] v = new int[1000001];
    static int[] b = new int[1000001];
    static int[] d = new int[1000001];

    static void dfs(int s, int k) {
        v[s] = 1;

        if (c[b[s]].size() >= k) {
            int size = c[b[s]].size();
            d[s] = c[b[s]].get(size - k);
        } else {
            d[s] = -1;
        }

        c[b[s]].add(s);

        for (int nei : a[s]) {
            if (v[nei] == 0) {
                dfs(nei, k);
            }
        }

        if (!c[b[s]].isEmpty()) {
            c[b[s]].remove(c[b[s]].size() - 1);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int k = sc.nextInt();

        for (int i = 1; i <= n; i++) {
            a[i] = new ArrayList<>();
            c[i] = new ArrayList<>();
        }

        for (int i = 1; i <= n; i++) {
            b[i] = sc.nextInt();
        }

        for (int i = 1; i < n; i++) {
            int x = sc.nextInt();
            int y = sc.nextInt();
            a[x].add(y);
            a[y].add(x);
        }

        dfs(1, k);

        for (int i = 1; i <= n; i++) {
            System.out.print(d[i] + " ");
        }
    }
}
